<a href="https://colab.research.google.com/github/Sebastian4524/Datavisualization/blob/main/pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **1. Library**

In [56]:
import numpy as np
import pandas as pd

# **2. Import a dataset**

In [57]:
s = "/datosp.xlsx"
d = pd.read_excel(s)

# **3. Exploration**

In [58]:
d.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 759 entries, 0 to 758
Data columns (total 16 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   Region                                        759 non-null    object 
 1   Country code                                  759 non-null    object 
 2   Economy                                       759 non-null    object 
 3   Reporting year                                759 non-null    int64  
 4   Survey name                                   759 non-null    object 
 5   Survey year                                   759 non-null    int64  
 6   Survey coverage                               755 non-null    object 
 7   Welfare type                                  755 non-null    object 
 8   Survey comparability                          755 non-null    float64
 9   Monetary (%)                                  759 non-null    flo

In [59]:
d.head()

Region Country code  Economy  Reporting year Survey name  Survey year  \
0    SSA          AGO   Angola            2018       IDREA         2018   
1    ECA          ALB  Albania            2012        LSMS         2012   
2    ECA          ALB  Albania            2016         HBS         2016   
3    ECA          ALB  Albania            2017         HBS         2017   
4    ECA          ALB  Albania            2018         HBS         2018   

  Survey coverage Welfare type  Survey comparability  Monetary (%)  \
0               N            c                 2.000        31.122   
1               N            c                 1.000         0.621   
2               N            c                 2.000         0.139   
3               N            c                 2.000         0.393   
4               N            c                 3.000         0.048   

   Educational attainment (%) Educational enrollment (%) Electricity (%)  \
0                      29.753                     27.443          52.640   
1                       2.271                      6.731           0.497   
2                       0.622                          -           0.431   
3                       0.331                          -           0.131   
4                       0.192                          -           0.060   

  Sanitation (%) Drinking water (%)  \
0         53.638             32.107   
1          3.040              0.343   
2          7.603              8.403   
3          6.976              9.270   
4          6.580              9.595   

   Multidimensional poverty headcount ratio (%)  
0                                        47.204  
1                                         0.793  
2                                         0.781  
3                                         0.711  
4                                         0.293

In [60]:
d.tail()

Region Country code       Economy  Reporting year Survey name  \
754    MNA          YEM   Yemen, Rep.            2014         HBS   
755    SSA          ZAF  South Africa            2014         LCS   
756    SSA          ZMB        Zambia            2015    LCMS-VII   
757    SSA          ZWE      Zimbabwe            2017       PICES   
758    SSA          ZWE      Zimbabwe            2019       PICES   

     Survey year Survey coverage Welfare type  Survey comparability  \
754         2014               N            c                 1.000   
755         2014               N            c                 3.000   
756         2015               N            c                 5.000   
757         2017               N            c                 0.000   
758         2019               N            c                 0.000   

     Monetary (%)  Educational attainment (%) Educational enrollment (%)  \
754        19.803                      15.951                     44.500   
755        20.501                       2.263                      2.339   
756        60.793                      24.375                     30.369   
757        34.409                       6.893                      3.101   
758        39.755                       0.927                      5.984   

    Electricity (%) Sanitation (%) Drinking water (%)  \
754          33.891         41.206             14.017   
755           4.094         35.171             10.361   
756          69.214         59.977             34.369   
757          37.062         67.637             22.405   
758          37.995         38.285             19.260   

     Multidimensional poverty headcount ratio (%)  
754                                        35.412  
755                                        21.710  
756                                        66.403  
757                                        39.898  
758                                        42.398

# **4. Cleaning**

## 4.1. Column names

In [61]:
d.rename(columns = {"Reporting year":"year"}, inplace = True)
d.rename(columns = {"Region":"region"}, inplace = True)
d.rename(columns = {"Survey comparability":"survey comparability"}, inplace = True)
d.rename(columns = {"Monetary (%)":"monetary"}, inplace = True)
d.rename(columns = {"Educational attainment (%)":"ea"}, inplace = True)
d.rename(columns = {"Educational enrollment (%)":"ee"}, inplace = True)
d.rename(columns = {"Electricity (%)":"electricity"}, inplace = True)
d.rename(columns = {"Sanitation (%)":"sanitation"}, inplace = True)
d.rename(columns = {"Drinking water (%)":"dwater"}, inplace = True)
d.rename(columns = {"Multidimensional poverty headcount ratio (%)":"multidimensional ratio"}, inplace = True)

In [62]:
d2=d

## 4.2. Delete columns

In [63]:
d2 = d2.drop(columns = ["Survey coverage","Welfare type","Survey name","survey comparability"])

In [64]:
d2.head()

region Country code  Economy  year  Survey year  monetary     ea     ee  \
0    SSA          AGO   Angola  2018         2018    31.122 29.753 27.443   
1    ECA          ALB  Albania  2012         2012     0.621  2.271  6.731   
2    ECA          ALB  Albania  2016         2016     0.139  0.622      -   
3    ECA          ALB  Albania  2017         2017     0.393  0.331      -   
4    ECA          ALB  Albania  2018         2018     0.048  0.192      -   

  electricity sanitation dwater  multidimensional ratio  
0      52.640     53.638 32.107                  47.204  
1       0.497      3.040  0.343                   0.793  
2       0.431      7.603  8.403                   0.781  
3       0.131      6.976  9.270                   0.711  
4       0.060      6.580  9.595                   0.293

## 4.3. Drop NA

In [65]:
d2 = d2.replace("-",np.NaN)
d2.head()

region Country code  Economy  year  Survey year  monetary     ea     ee  \
0    SSA          AGO   Angola  2018         2018    31.122 29.753 27.443   
1    ECA          ALB  Albania  2012         2012     0.621  2.271  6.731   
2    ECA          ALB  Albania  2016         2016     0.139  0.622    NaN   
3    ECA          ALB  Albania  2017         2017     0.393  0.331    NaN   
4    ECA          ALB  Albania  2018         2018     0.048  0.192    NaN   

   electricity  sanitation  dwater  multidimensional ratio  
0       52.640      53.638  32.107                  47.204  
1        0.497       3.040   0.343                   0.793  
2        0.431       7.603   8.403                   0.781  
3        0.131       6.976   9.270                   0.711  
4        0.060       6.580   9.595                   0.293

In [66]:
d2 =d2.dropna()

In [67]:
d2.head()

region Country code    Economy  year  Survey year  monetary     ea     ee  \
0    SSA          AGO     Angola  2018         2018    31.122 29.753 27.443   
1    ECA          ALB    Albania  2012         2012     0.621  2.271  6.731   
5    LAC          ARG  Argentina  2010         2010     0.749  1.904  0.964   
6    LAC          ARG  Argentina  2011         2011     0.707  1.808  1.051   
7    LAC          ARG  Argentina  2012         2012     0.641  1.650  0.834   

   electricity  sanitation  dwater  multidimensional ratio  
0       52.640      53.638  32.107                  47.204  
1        0.497       3.040   0.343                   0.793  
5        0.000       0.683   0.326                   0.772  
6        0.000       0.717   0.309                   0.744  
7        0.000       0.524   0.324                   0.675

In [68]:
d2["year_str"] = d2 ["year"].astype(str)

<ipython-input-68-f047fb21a0aa>:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



# **5. Indexing**

In [69]:
d2 = d2.set_index(["Economy","year"])
d2.head()

region Country code  Survey year  monetary     ea     ee  \
Economy   year                                                            
Angola    2018    SSA          AGO         2018    31.122 29.753 27.443   
Albania   2012    ECA          ALB         2012     0.621  2.271  6.731   
Argentina 2010    LAC          ARG         2010     0.749  1.904  0.964   
          2011    LAC          ARG         2011     0.707  1.808  1.051   
          2012    LAC          ARG         2012     0.641  1.650  0.834   

                electricity  sanitation  dwater  multidimensional ratio  \
Economy   year                                                            
Angola    2018       52.640      53.638  32.107                  47.204   
Albania   2012        0.497       3.040   0.343                   0.793   
Argentina 2010        0.000       0.683   0.326                   0.772   
          2011        0.000       0.717   0.309                   0.744   
          2012        0.000       0.524   0.324                   0.675   

               year_str  
Economy   year           
Angola    2018     2018  
Albania   2012     2012  
Argentina 2010     2010  
          2011     2011  
          2012     2012

# **6. Descriptive Statistics**

In [70]:
# Basic
d2[["ea","ee","electricity","sanitation","dwater","multidimensional ratio"]].describe()

ea      ee  electricity  sanitation  dwater  multidimensional ratio
count 336.000 336.000      336.000     336.000 336.000                 336.000
mean   13.554   6.946       12.455      20.977   9.781                  14.225
std    15.987  10.373       22.527      24.115  11.668                  20.938
min     0.000   0.055        0.000       0.000   0.000                   0.000
25%     3.225   1.297        0.232       3.212   1.562                   1.328
50%     6.302   2.846        1.600      11.082   5.691                   5.093
75%    17.103   6.312        9.914      29.749  12.327                  16.907
max    75.149  58.043       97.251      95.898  69.207                  85.678

In [71]:
import pandas as pd

# Define una función personalizada para el formato
def custom_format(x):
    if isinstance(x, float):
        return '{:.3f}'.format(x)
    return str(x)

# Aplica la función de formato personalizada a todos los datos del DataFrame
pd.options.display.float_format = custom_format

# Ejecuta tu código
d2[["ea", "ee", "electricity", "sanitation", "dwater", "multidimensional ratio"]].describe()

ea      ee  electricity  sanitation  dwater  multidimensional ratio
count 336.000 336.000      336.000     336.000 336.000                 336.000
mean   13.554   6.946       12.455      20.977   9.781                  14.225
std    15.987  10.373       22.527      24.115  11.668                  20.938
min     0.000   0.055        0.000       0.000   0.000                   0.000
25%     3.225   1.297        0.232       3.212   1.562                   1.328
50%     6.302   2.846        1.600      11.082   5.691                   5.093
75%    17.103   6.312        9.914      29.749  12.327                  16.907
max    75.149  58.043       97.251      95.898  69.207                  85.678

In [72]:
d2[["ea", "ee" , "sanitation", "dwater", "multidimensional ratio"]].describe().applymap(lambda x:"{:,.1f}".format(x))

ea     ee sanitation dwater multidimensional ratio
count  336.0  336.0      336.0  336.0                  336.0
mean    13.6    6.9       21.0    9.8                   14.2
std     16.0   10.4       24.1   11.7                   20.9
min      0.0    0.1        0.0    0.0                    0.0
25%      3.2    1.3        3.2    1.6                    1.3
50%      6.3    2.8       11.1    5.7                    5.1
75%     17.1    6.3       29.7   12.3                   16.9
max     75.1   58.0       95.9   69.2                   85.7

# **7. Conclusions**

<ul>
  <li>The country that has the highest educational attainment is 75.1 % meaning that just 24.9% of their population does not have a degree</li>
  <li>the ceros on the descriptive data might be that the country could be so poor that they can not measure their sanitation or drinking water </li>
  <li>We can see that the highest multidimentional ratio is a combination of the other high variables </li>
</ul>


In [73]:
d2[["dwater"]].groupby("year").describe().applymap(lambda x:"{:,.1f}".format(x))

dwater                                          
      count  mean   std   min   25%   50%   75%   max
year                                                 
2009    1.0  69.2   nan  69.2  69.2  69.2  69.2  69.2
2010   25.0  11.9  13.1   0.2   2.4   7.8  17.3  50.6
2011   21.0  12.8  14.2   0.1   2.0   7.7  18.2  56.0
2012   31.0  14.6  15.8   0.3   2.6   8.7  22.6  59.9
2013   23.0   7.0   7.4   0.3   1.2   5.2   8.7  25.6
2014   26.0  13.1  12.6   0.4   4.5   9.2  20.0  44.9
2015   34.0  11.2  12.3   0.0   2.1   8.0  11.0  42.7
2016   33.0   9.6  10.9   0.0   1.6   4.7  12.8  40.8
2017   29.0   5.9   6.4   0.1   0.7   3.9   8.7  22.4
2018   42.0  11.9  11.6   0.1   2.0   7.3  20.9  37.5
2019   31.0   5.9   6.5   0.0   0.7   3.1   9.0  23.7
2020   16.0   3.9   2.7   0.1   1.6   4.3   5.7   8.6
2021   17.0   3.9   3.7   0.3   0.7   3.7   5.3  13.6
2022    7.0   2.6   2.3   0.0   0.8   2.6   4.0   6.0

# **8. Conclusions**

<ul>
  <li>with the pass of de tiempo we can see tht in the world the drinking water in decreasing for everyone</li>
  <li>the year in which we were able to see the highest metric for the cuantity of people that have access to drinking wated was in 2009 </li>
  <li>we can see that the worst year for the drinking water metric was in 2022  </li>
</ul>

In [74]:
# by country
Eco = d2.filter(like = "Argentina", axis = 0)
Argentina.head()

region Country code    Economy  year  Survey year  monetary    ea    ee  \
5    LAC          ARG  Argentina  2010         2010     0.749 1.904 0.964   
6    LAC          ARG  Argentina  2011         2011     0.707 1.808 1.051   
7    LAC          ARG  Argentina  2012         2012     0.641 1.650 0.834   
8    LAC          ARG  Argentina  2013         2013     0.623 1.619 0.896   
9    LAC          ARG  Argentina  2014         2014     0.612 1.418 0.949   

   electricity  sanitation  dwater  multidimensional ratio  
5        0.000       0.683   0.326                   0.772  
6        0.000       0.717   0.309                   0.744  
7        0.000       0.524   0.324                   0.675  
8        0.000       0.406   0.321                   0.657  
9        0.000       0.399   0.406                   0.635

In [75]:
# by year
y2009 =d2.query("year == 2009")
y2009.head()

region Country code  Survey year  monetary     ea    ee  \
Economy          year                                                           
Papua New Guinea 2009    EAP          PNG         2009    39.683 22.216 9.004   

                       electricity  sanitation  dwater  \
Economy          year                                    
Papua New Guinea 2009       82.574      79.779  69.207   

                       multidimensional ratio year_str  
Economy          year                                   
Papua New Guinea 2009                  74.658     2009

# **9. Pivot**


In [76]:
## Be careful you must use the un indexed dataframe

## exaple
## rows = year
## columns = country code
## values = pop_g

pivot1 = d.pivot(index = "year",
                 columns = "Economy",
                 values = "ea")
pivot1.head()

Economy  Albania  Angola  Argentina  Armenia  Australia  Austria  Bangladesh  \
year                                                                           
2009         NaN     NaN        NaN      NaN        NaN    0.550         NaN   
2010         NaN     NaN      1.904      NaN      2.600    0.295      29.144   
2011         NaN     NaN      1.808      NaN        NaN    0.296         NaN   
2012       2.271     NaN      1.650      NaN        NaN    0.472         NaN   
2013         NaN     NaN      1.619      NaN        NaN    0.000         NaN   

Economy  Belarus  Belgium  Benin  ...  Uganda  Ukraine  United States  \
year                              ...                                   
2009         NaN    1.868    NaN  ...     NaN      NaN            NaN   
2010       0.024    1.358    NaN  ...     NaN      NaN          0.324   
2011       0.062    1.345    NaN  ...     NaN      NaN          0.277   
2012       0.028    0.914    NaN  ...  47.914      NaN          0.298   
2013       0.013    1.407    NaN  ...     NaN      NaN          0.302   

Economy  Uruguay  Vanuatu  Viet Nam  West Bank and Gaza  Yemen, Rep.  Zambia  \
year                                                                           
2009         NaN      NaN       NaN                 NaN          NaN     NaN   
2010       4.388   18.015     6.667               3.764          NaN     NaN   
2011       3.506      NaN       NaN               3.226          NaN     NaN   
2012       3.119      NaN     6.642                 NaN          NaN     NaN   
2013       3.022      NaN       NaN                 NaN          NaN     NaN   

Economy  Zimbabwe  
year               
2009          NaN  
2010          NaN  
2011          NaN  
2012          NaN  
2013          NaN  

[5 rows x 149 columns]

# **10. Graphs**
*texto en cursiva*

In [77]:
import plotly.express as px

# **10.1 Line Graph**


In [78]:
Argentina = d2.query("Economy == 'Argentina'")
Argentina.head()

region Country code  Survey year  monetary    ea    ee  \
Economy   year                                                          
Argentina 2010    LAC          ARG         2010     0.749 1.904 0.964   
          2011    LAC          ARG         2011     0.707 1.808 1.051   
          2012    LAC          ARG         2012     0.641 1.650 0.834   
          2013    LAC          ARG         2013     0.623 1.619 0.896   
          2014    LAC          ARG         2014     0.612 1.418 0.949   

                electricity  sanitation  dwater  multidimensional ratio  \
Economy   year                                                            
Argentina 2010        0.000       0.683   0.326                   0.772   
          2011        0.000       0.717   0.309                   0.744   
          2012        0.000       0.524   0.324                   0.675   
          2013        0.000       0.406   0.321                   0.657   
          2014        0.000       0.399   0.406                   0.635   

               year_str  
Economy   year           
Argentina 2010     2010  
          2011     2011  
          2012     2012  
          2013     2013  
          2014     2014

In [81]:
# fig 1
fig1 = px.line(Argentina, x ="year_str", y = "multidimensional ratio",
               title="my beautiful grafic")
fig1.show()

In [95]:
# fig 1
fig2 = px.line(Argentina, x ="year_str", y = "multidimensional ratio")

#title
fig2.update_layout(title_text = " Beautiful grafic ",
                   title_font_size  = 20,
                   title_x = 0.5)
# axis
fig2.update_xaxes(title = "year", title_font_size = 20)
fig2.update_yaxes(title = "Multidimensional ratio", title_font_size = 20)
fig2.show()